core/iwasm/common/wasm_application.c

```diff
@@ -96,22 +92,25 @@ wasm_application_execute_main(WASMModuleInstanceCommon *module_inst, int32 argc,
     bool ret, is_import_func = true;
 
 #if WASM_ENABLE_LIBC_WASI != 0
-    if (wasm_runtime_is_wasi_mode(module_inst)) {
-        /* In wasi mode, we should call function named "_start"
-           which initializes the wasi envrionment and then calls
-           the actual main function. Directly call main function
-           may cause exception thrown. */
-        if ((func = wasm_runtime_lookup_wasi_start_function(module_inst)))
-            return wasm_runtime_create_exec_env_and_call_wasm(module_inst, func,
-                                                              0, NULL);
-        /* If no start function was found, we execute
-           the main function as normal */
+    /* In wasi mode, we should call function named "_start"
+       which initializes the wasi envrionment and then calls
+       the actual main function. Directly call main function
+       may cause exception thrown. */
+    if ((func = wasm_runtime_lookup_wasi_start_function(module_inst))) {
+        return wasm_runtime_create_exec_env_and_call_wasm(module_inst, func, 0,
+                                                          NULL);
+    }
+    else {
+        wasm_runtime_set_exception(
```

Comment:
Should not set exception here: (1) Must the wasm app compiled as wasi mode export _start function? (2) Here the WASM_ENABLE_LIBC_WASI is enabled, but it doesn't require that the wasm app must be a wasi mode app, the app can be non-wasi mode also, for example `/opt/wasi-sdk/bin/clang -nostdlib --no-entry --export=__main_argc_argv ...`

Reply:
If with WASI, a .wasm is either a command or a reactor. There should be not a third case. So. YES we run the WASM app if there is a `_start`.  And we quit with a message if there is no `_start`.

Reply:
Plus, we shall treat `-nostdlib` as a behavior of disabled WASI.